### PR TITLE
Respecting required  field of inner case classes in Coder macro

### PR DESCRIPTION
**Pull request by shnapz**
Fixing [bug #4643](https://github.com/spotify/scio/issues/4643):
- When Scala case class is compiled to Java byte code it contains a required synthetic field `$outer` as a reference to the instance where it is created;
- Coder Magnolia macro emits anonymous implementation of `magnolia1.CaseClass[Typeclass[_], Type]` that:
  - has a constructor that accepts a reference to `$outer`
  - creates a case class via `rawConstruct` method, passing a reference to `$outer`
  - private class CaseClassConstructor is intended to be serializable factory of `magnolia1.CaseClass` implementation
**The problem** was that `CaseClassConstructor` created `magnolia1.CaseClass` using:
`ClosureCleaner.instantiateClass(cls)`
That created an instance without constructor and left `$outer` field null. If `$outer` **is null Scala case class constructor throws exception**.

**The fix:**
- In short: instantiate empty instance of `$outer` object and create an instance of `magnolia1.CaseClass[Typeclass[_], Type]` using constructor.
- Longer story: there could be multiple synthetic classes generated by a compiler wrapping Scala case class (they have only one constructor that accepts `$outer`), so the recursion (Trampoline-d) is used to reach the first package-level class

### CODE REVIEW

**Comment by shnapz** [comment]
@RustedBones please re-check the PR

**Review by RustedBones — APPROVED** [review]

File: scio-core/src/main/scala/com/spotify/scio/coders/DerivedCoder.scala [hunk 1]
```diff
@@ -41,12 +41,32 @@ object LowPriorityCoderDerivation {
...
+         another class. And the constructor of that compiled class requires outer field to be not
+          null.
+         If "$outer" is present it's an inner class and this scenario is officially not supported
+          by Scio */
+        case Some(_) =>
+          throw new Throwable(
+            s"Found an $$outer field in $ctxClass. Possibly it is an attempt to use inner case " +
+              s"class in a Scio transformation. Inner case classes are not supported in Scio " +
+              s"auto-derived macros. Move the case class to the package level or define a custom " +
+              s"coder."
```
[the remark this answers is not in the document]
Comment (shnapz) [reply scio-core/src/main/scala/com/spotify/scio/coders/DerivedCoder.scala]: Done!
